### PR TITLE
Add Targets

### PR DESCRIPTION
**Pull request by peter7z**
Adds Target model, create method in controller, and some tests.

### CODE REVIEW

File: app/models/target.rb [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+#
+
+class Target < ApplicationRecord
+  belongs_to :topic
+  belongs_to :user
+
+  validates :title, presence: true
+  validates :radius, presence: true, numericality: true
+  validates :latitude, presence: true, numericality: true
+  validates :longitude, presence: true, numericality: true
```
Comment (TimoPeraza) [inline app/models/target.rb]: This validates could be in one sentece like: ```validates :latitude, :longitude, presence: true, numericality: true```

File: app/models/target.rb [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+#  index_targets_on_topic_id  (topic_id)
+#  index_targets_on_user_id   (user_id)
+#
+
+class Target < ApplicationRecord
+  belongs_to :topic
+  belongs_to :user
+
+  validates :title, presence: true
+  validates :radius, presence: true, numericality: true
```
Comment (TimoPeraza) [inline app/models/target.rb]: Maybe you have to validate if the radius is greater/less than a fair value

File: app/models/user.rb [hunk 3]
```diff
@@ -37,4 +37,6 @@ class User < ActiveRecord::Base
   devise :database_authenticatable, :registerable,
          :recoverable, :rememberable, :validatable
   include DeviseTokenAuth::Concerns::User
+
+  has_many :targets
```
Comment (TimoPeraza) [inline app/models/user.rb]: maybe could add a ```dependent: :destroy```

File: spec/factories/user.rb [hunk 4]
```diff
@@ -0,0 +1,8 @@
+require 'faker'
+
+FactoryBot.define do
+  factory :user do
+    email { Faker::Internet.email }
+    password { Faker::Internet.password }
+  end
+end
```
Comment (TimoPeraza) [inline spec/factories/user.rb]: missing newline 😄

File: app/controllers/targets_controller.rb [hunk 5]
```diff
@@ -0,0 +1,15 @@
+class TargetsController < ApplicationController
+  before_action :authenticate_user!
+
+  def create
+    current_user.targets.create!(target_params)
+  rescue ActiveRecord::RecordInvalid => e
```
Comment (t-romani) [inline app/controllers/targets_controller.rb]: Maybe you can include this on ApplicationController instead using something like `rescue_from`

File: spec/factories/user.rb [hunk 6]
```diff
@@ -0,0 +1,8 @@
+require 'faker'
+
+FactoryBot.define do
+  factory :user do
+    email { Faker::Internet.email }
+    password { Faker::Internet.password }
```
Comment (t-romani) [inline spec/factories/user.rb]: You can align assignments

File: spec/models/target_spec.rb [hunk 7]
```diff
@@ -0,0 +1,11 @@
+require 'rails_helper'
+
+describe Target do
+  describe 'validations' do
+    subject { build :target }
+    it { is_expected.to validate_presence_of(:title) }
+    it { is_expected.to validate_presence_of(:radius) }
```
Comment (TimoPeraza) [inline spec/models/target_spec.rb]: You should add validations for ```greater_than``` and ```less_or_equal_than```

File: spec/factories/target.rb [hunk 8]
```diff
@@ -0,0 +1,11 @@
...
+
+FactoryBot.define do
+  factory :target do
+    topic
+    title     { Faker::Cannabis.buzzword }
+    latitude  { Faker::Number.decimal(l_digits: 2, r_digits: 2) }
+    longitude { Faker::Number.decimal(l_digits: 2, r_digits: 2) }
+    radius    { Faker::Number.between(from: 1, to: 5000) }
+  end
+end
```
Comment (martinjaimem) [inline spec/factories/target.rb]: Add new line at eof 😄

File: spec/models/target_spec.rb [hunk 9]
```diff
@@ -0,0 +1,11 @@
+require 'rails_helper'
+
+describe Target do
+  describe 'validations' do
+    subject { build :target }
+    it { is_expected.to validate_presence_of(:title) }
```
Comment (martinjaimem) [inline spec/models/target_spec.rb]: Leave a blank line between `subject` and `it {...`

File: spec/models/target_spec.rb [hunk 10]
```diff
@@ -0,0 +1,11 @@
+require 'rails_helper'
+
+describe Target do
```
Comment (martinjaimem) [inline spec/models/target_spec.rb]: Add `, :type => :model` here.

https://typefast.blog/custom-spec-types-with-rspec/

File: spec/requests/targets/create_spec.rb [hunk 11]
```diff
@@ -0,0 +1,46 @@
...
+  end
+
+  subject { post targets_path, params: params, headers: auth_user_headers, as: :json }
+
+  it 'returns a successful response' do
+    subject
+    expect(response).to be_successful
+  end
+
+  it 'creates the target' do
```
Comment (martinjaimem) [inline spec/requests/targets/create_spec.rb]: I would also test that the created target attribute values match the ones in the params, WDYT?

File: app/controllers/application_controller.rb [hunk 12]
```diff
@@ -1,6 +1,48 @@
 class ApplicationController < ActionController::API
+  include DeviseTokenAuth::Concerns::SetUserByToken
   before_action :configure_permitted_parameters, if: :devise_controller?
 
+  rescue_from Exception,                           with: :render_error
```
Comment (martinjaimem) [inline app/controllers/application_controller.rb]: Nice 🚀

File: app/controllers/targets_controller.rb [hunk 13]
```diff
@@ -0,0 +1,13 @@
+class TargetsController < ApplicationController
+  before_action :authenticate_user!
```
Comment (gabrielbursztein2) [inline app/controllers/targets_controller.rb]: In the future, you will need to authenticate the user in more than one controller. Create an ApiControllerUser and inherit from it.

File: app/models/target.rb [hunk 14]
```diff
@@ -0,0 +1,28 @@
...
+#  index_targets_on_user_id   (user_id)
+#
+
+class Target < ApplicationRecord
+  belongs_to :topic
+  belongs_to :user
+
+  validates :title, presence: true
+  validates :radius, presence: true, numericality: { greater_than: 0, less_or_equal_than: 5000 }
+  validates :latitude, :longitude, presence: true, numericality: { greater_than: 0 }
```
Comment (gabrielbursztein2) [inline app/models/target.rb]: Latitude and longitude could be negative, remove the constraint.

File: spec/factories/target.rb [hunk 15]
```diff
@@ -0,0 +1,11 @@
+require 'faker'
+
+FactoryBot.define do
+  factory :target do
+    topic
+    title     { Faker::Cannabis.buzzword }
+    latitude  { Faker::Number.decimal(l_digits: 2, r_digits: 2) }
```
Comment (gabrielbursztein2) [inline spec/factories/target.rb]: `Faker::Address.latitude`. Same below

File: spec/requests/targets/create_spec.rb [hunk 16]
```diff
@@ -0,0 +1,46 @@
...
+describe 'POST /targets', type: :request do
+
+  let!(:user)  { create(:user) }
+  let!(:topic) { create(:topic) }
+
+  let(:params) do
+    {
+      title:     Faker::Cannabis.buzzword,
+      topic_id:  topic.id,
+      latitude:  Faker::Number.decimal(l_digits: 2, r_digits: 2),
```
Comment (gabrielbursztein2) [inline spec/requests/targets/create_spec.rb]: Same here

File: spec/requests/targets/create_spec.rb [hunk 17]
```diff
@@ -0,0 +1,46 @@
...
+
+  it 'returns a successful response' do
+    subject
+    expect(response).to be_successful
+  end
+
+  it 'creates the target' do
+    expect {
+      subject
+    }.to change(Target, :count).by(1)
```
Comment (gabrielbursztein2) [inline spec/requests/targets/create_spec.rb]: Write it in one line when possible.

**Pull request merged**